### PR TITLE
[ListView] Call 'props.onEndReached' each time the scroll offset crosses the threshold near the end

### PR DESCRIPTION
**Pull request by aleclarson**
This is in response to #1967. The proposition is that `onEndReached` should work as expected and get called every time the `onEndReachedThreshold` is passed. Currently, it's only called once unless the `contentLength` is set to a new value.

### CODE REVIEW

File: Libraries/CustomComponents/ListView/ListView.js [hunk 1]
```diff
@@ -484,9 +484,8 @@ var ListView = React.createClass({
   },
 
   _getDistanceFromEnd: function(scrollProperties) {
-    return scrollProperties.contentLength -
-      scrollProperties.visibleLength -
-      scrollProperties.offset;
+    var maxHeight = Math.max(scrollProperties.contentLength, scrollProperties.visibleLength);
```
Comment (aleclarson) [inline Libraries/CustomComponents/ListView/ListView.js]: In case the `dataSource` is empty, the `contentLength` is potentially less than the `visibleLength`.

File: Libraries/CustomComponents/ListView/ListView.js [hunk 2]
```diff
@@ -559,28 +558,32 @@ var ListView = React.createClass({
...
+
+        this._isEndReached = isEndReached;
+
+        if (isEndReached) {
+          this.props.onEndReached(e);
+        }
+      }
     }
 
+    this._renderMoreRowsIfNeeded();
```
Comment (aleclarson) [inline Libraries/CustomComponents/ListView/ListView.js]: Removed the comparison between `this.state.curRenderedRowsCount` and `this.props.dataSource.getRowCount()` since `this._renderMoreRowsIfNeeded()` already does the same comparison.

File: Libraries/CustomComponents/ListView/ListView.js (hunk 1 above)
Comment (aleclarson) [inline Libraries/CustomComponents/ListView/ListView.js]: Maybe this should be in its own commit?

Comment (ide) [inline Libraries/CustomComponents/ListView/ListView.js]: Separate PR sounds good since it's easier to go through the commit logs and bisect if there's an issue. Could you rename `maxHeight` to `maxLength` since the ScrollView could be horizontal, and wrap the code to fit 80 cols?

File: Libraries/CustomComponents/ListView/ListView.js [hunk 3]
```diff
@@ -559,28 +558,32 @@ var ListView = React.createClass({
...
-    } else {
-      this._renderMoreRowsIfNeeded();
+    var lengthKey = this.props.horizontal ? 'width' : 'height';
+    var offsetKey = this.props.horizontal ? 'x' : 'y';
+
+    this.scrollProperties.visibleLength = e.nativeEvent.layoutMeasurement[lengthKey];
+    this.scrollProperties.contentLength = e.nativeEvent.contentSize[lengthKey];
+    this.scrollProperties.offset = e.nativeEvent.contentOffset[offsetKey];
+
+    this._updateVisibleRows(e);
```
Comment (ide) [inline Libraries/CustomComponents/ListView/ListView.js]: `e.nativeEvent.updatedChildFrames`

File: Libraries/CustomComponents/ListView/ListView.js [hunk 4]
```diff
@@ -559,28 +558,32 @@ var ListView = React.createClass({
...
+
+    if (this.props.onEndReached) {
+
+      var distanceFromEnd = this._getDistanceFromEnd(this.scrollProperties);
+      var isEndReached = distanceFromEnd < this.props.onEndReachedThreshold;
+
+      if (isEndReached !== this._isEndReached) {
+
+        this._isEndReached = isEndReached;
```
Comment (ide) [inline Libraries/CustomComponents/ListView/ListView.js]: I think I agree with the attached issue... if you scroll to the end, then scroll back up, and then scroll down again, you should get another "onEndReached" event. This gives the owner more flexibility over whether it wants to dedupe the event or not.

File: Libraries/CustomComponents/ListView/ListView.js [hunk 5]
```diff
@@ -559,28 +558,32 @@ var ListView = React.createClass({
...
+    var offsetKey = this.props.horizontal ? 'x' : 'y';
+
+    this.scrollProperties.visibleLength = e.nativeEvent.layoutMeasurement[lengthKey];
+    this.scrollProperties.contentLength = e.nativeEvent.contentSize[lengthKey];
+    this.scrollProperties.offset = e.nativeEvent.contentOffset[offsetKey];
+
+    this._updateVisibleRows(e);
+
+    if (this.props.onEndReached) {
+
```
Comment (ide) [inline Libraries/CustomComponents/ListView/ListView.js]: style: remove newline

File: Libraries/CustomComponents/ListView/ListView.js [hunk 6]
```diff
@@ -559,28 +558,32 @@ var ListView = React.createClass({
...
+      var distanceFromEnd = this._getDistanceFromEnd(this.scrollProperties);
+      var isEndReached = distanceFromEnd < this.props.onEndReachedThreshold;
+
+      if (isEndReached !== this._isEndReached) {
+
```
Comment (ide) [inline Libraries/CustomComponents/ListView/ListView.js]: here too

**Comment by ide** [comment]
This looks good to me at a high level.

@sahrens adding you to this PR to have a look and import it if it looks good after my comments have been addressed.

**Comment by aleclarson** [comment]
Everything is updated. :+1:

**Comment by paramaggarwal** [comment]
Looks good, let's merge this in please!

**Comment by sahrens** [comment]
I definitely have some concerns here.  We've had a lot of subtle bugs with the exact behavior here - onEndReach not getting called or getting called too many times in all sorts of different scenarios like the content being shorter than the screen etc - so we need to be really careful.

For example, with the PR you have here, what if more content is loaded, but it only adds a small amount of content to the total height such that you're still within the `onEndReachedThreshold` range?  Then you won't be able to trigger another `onEndReached` without the user scrolling back up them down again.

Perhaps we should track both `_sentEndForContentLength` and `_isEndReached` so we can trigger another `onEndReached` call in either case?

I'm also not sure you should remove the check for `this.state.curRenderedRowsCount === this.props.dataSource.getRowCount()` - without that check, I think it's possible for onEndReached to render several times, once for each row that is incrementally rendered per frame, which probably isn't what the caller wants.

**Comment by grubstarstar** [comment]
how do we work around this?
